### PR TITLE
Improvement/arsn 157 short i ds

### PR DESCRIPTION
**Pull request by watsaqat**

### CODE REVIEW

File: lib/versioning/VersionID.js [hunk 1]
```diff
@@ -116,12 +116,17 @@ function generateVersionId(info, replicationGroupId) {
     lastSeq = (lastTimestamp === ts) ? lastSeq + 1 : 0;
     lastTimestamp = ts;
 
+    // if USE_HEX_VERSION_ID is set, info is used. By default, it is not used.
+    if (process.env.USE_HEX_VERSION_ID === 'no' || !process.env.USE_HEX_VERSION_ID) {
```
Comment (watsaqat) [inline lib/versioning/VersionID.js]: better name for this?

File: lib/versioning/VersionID.js [hunk 2]
```diff
@@ -221,13 +226,13 @@ const ENC_TYPE_BASE62 = 1; // new (tiny) encoding
...
  * @return {string} - the encoded versionId
  */
-function encode(str, encType = ENC_TYPE_HEX) {
-    if (encType === ENC_TYPE_BASE62) {
-        return base62Encode(str);
-    }
+function encode(str) {
+    // default format without 'info' field will always be 27 characters
+    if (str.length === 27) {
+        return padRight(base62Encode(str), ' '.repeat(32));
```
Comment (watsaqat) [inline lib/versioning/VersionID.js]: any potential issues with padding right with spaces?

File: tests/unit/versioning/VersionID.spec.js [hunk 3]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
+                }
+            }
+        });
 
-    it('error case: when invalid base62 key part 3 has invalid base62 character', () => {
-        const invalidBase62VersionId = 'aJLWKz4Ko9IjBBgXKj5KQT.G9UHv0g7P';
-        const decoded = VID.base62Decode(invalidBase62VersionId);
-        assert(decoded instanceof Error);
-    });
+        it('simple base62 version test', () => {
```
Comment (watsaqat) [inline tests/unit/versioning/VersionID.spec.js:73]: base62Encode/Decode and hexEncode/Decode don't depend on USE_HEX_VERSION_ID, so they should be moved to their own section as was previously done.

File: lib/versioning/VersionID.js (hunk 1 above)
Comment (jonathan-gramain) [reply lib/versioning/VersionID.js]: as discussed, I suggest checking for `process.env.S3_VERSION_ID_ENCODING_TYPE=hex` specifically

File: tests/unit/versioning/VersionID.spec.js [hunk 4]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
+function generateRandomVIDs(count) {
     const vids = Array(count).fill(null);
     for (let i = 0; i < count; i++) {
         vids[i] = VID.generateVersionId(randkey(15), 'PARIS');
     }
     process.env.VID_CRYPTO_PASSWORD = randkey(64);
+    return vids;
+}
+
+const count = 1000;
```
Comment (vrancurel) [inline tests/unit/versioning/VersionID.spec.js]: is it possible to have 1,000,000 ?

File: lib/versioning/VersionID.js [hunk 5]
```diff
@@ -202,8 +215,11 @@ function base62Decode(str) {
         const enc2 = str.substring(start, start + B62V_EPAD.length);
         const orig2 = base62Integer.decode(enc2);
         start += B62V_EPAD.length;
-        const enc3 = str.substring(start);
+        let enc3 = str.substring(start);
+        // take off null bytes padding
+        enc3 = enc3.replaceAll('\u0000', '');
```
Comment (vrancurel) [inline lib/versioning/VersionID.js]: is there a better way to trim the null bytes ?

**Review by vrancurel — CHANGES_REQUESTED** [review]

**Comment by vrancurel** [comment]
a few details but otherwise it looks good.

File: lib/versioning/VersionID.js [hunk 6]
```diff
@@ -202,8 +215,12 @@ function base62Decode(str) {
         const enc2 = str.substring(start, start + B62V_EPAD.length);
         const orig2 = base62Integer.decode(enc2);
         start += B62V_EPAD.length;
-        const enc3 = str.substring(start);
+        let enc3 = str.substring(start);
+        // take off null bytes padding only from the end
+        const re = new RegExp('\u0000+$'); // eslint-disable-line
```
Comment (vrancurel) [inline lib/versioning/VersionID.js]: isn't it more efficient to do:
`const i = b.indexOf(0x00);
  const nb = b.slice(0, i);`

**Review by vrancurel — CHANGES_REQUESTED** [review]

File: lib/versioning/VersionID.js (hunk 6 above)
Comment (watsaqat) [reply lib/versioning/VersionID.js]: as long as we are guaranteed to never have a null byte besides the padding at the end, we can do this. Benchmark is about 2x speedup in favor of slice

Comment (watsaqat) [reply lib/versioning/VersionID.js]: https://measurethat.net/Benchmarks/Show/18405/0/regex-vs-slice
I tried binary search too lol, but the strings are too short to take advantage of over a simple while loop

**Review by vrancurel — DISMISSED** [review]

File: lib/versioning/VersionID.js (hunk 6 above)
Comment (vrancurel) [reply lib/versioning/VersionID.js]: > as long as we are guaranteed to never have a null byte besides the padding at the end, we can do this. Benchmark is about 2x speedup in favor of slice

why would we have a null byte ?

File: lib/versioning/VersionID.js [hunk 7]
```diff
@@ -116,12 +116,19 @@ function generateVersionId(info, replicationGroupId) {
     lastSeq = (lastTimestamp === ts) ? lastSeq + 1 : 0;
     lastTimestamp = ts;
 
+    // if USE_HEX_VERSION_ID is set, info is used. By default, it is not used.
+    if (process.env.S3_VERSION_ID_ENCODING_TYPE === 'hex') {
+        // info field stays as is
+    } else {
+        info = ''; // eslint-disable-line
+    }
```
Comment (miniscruff) [inline lib/versioning/VersionID.js:125]:
```suggestion
    if (process.env.S3_VERSION_ID_ENCODING_TYPE !== 'hex') {
        info = ''; // eslint-disable-line
    }
```
Having an empty if looks weird to me, I would flip the condition

File: lib/versioning/VersionID.js [hunk 8]
```diff
@@ -202,8 +215,19 @@ function base62Decode(str) {
         const enc2 = str.substring(start, start + B62V_EPAD.length);
         const orig2 = base62Integer.decode(enc2);
         start += B62V_EPAD.length;
-        const enc3 = str.substring(start);
+        let enc3 = str.substring(start);
+        // take off null bytes padding only from the end
```
Comment (miniscruff) [inline lib/versioning/VersionID.js]: You can use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/lastIndexOf to find the last character you need. TrimEnd might work but doesn't seem like it unless \u0000 counts as white space.

File: tests/unit/versioning/VersionID.spec.js [hunk 9]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
+        const vids = generateRandomVIDs(count);
 
-    it('sorted in reversed chronological and alphabetical order', () => {
-        for (let i = 0; i < count; i++) {
-            if (i !== 0) {
-                assert(vids[i - 1] > vids[i],
-                    'previous VersionID is higher than its next');
+        it('sorted in reversed chronological and alphabetical order', () => {
+            for (let i = 0; i < count; i++) {
+                if (i !== 0) {
```
Comment (miniscruff) [inline tests/unit/versioning/VersionID.spec.js]: You can just start at 1 to avoid this extra check

File: tests/unit/versioning/VersionID.spec.js [hunk 10]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
-    });
+        });
 
-    // nodejs 10 no longer returns error for non-hex string versionIds
-    it.skip('should return error decoding non-hex string versionIds', () => {
-        const encoded = vids.map(vid => VID.hexEncode(vid));
-        const decoded = encoded.map(vid => VID.hexDecode(`${vid}foo`));
-        decoded.forEach(result => assert(result instanceof Error));
-    });
+        // nodejs 10 no longer returns error for non-hex string versionIds
```
Comment (miniscruff) [inline tests/unit/versioning/VersionID.spec.js:37]: Can probably remove this since it was a node 10 thing and we are on node 16 now

File: tests/unit/versioning/VersionID.spec.js [hunk 11]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
-    it('simple base62 version test', () => {
-        const vid = '98376906954349999999RG001  145.20.5';
-        const encoded = VID.base62Encode(vid);
-        assert.strictEqual(encoded, 'aJLWKz4Ko9IjBBgXKj5KQT2G9UHv0g7P');
-        const decoded = VID.base62Decode(encoded);
-        assert.strictEqual(vid, decoded);
-    });
+        it('should encode and decode correctly with legacy format', () => {
+            const encoded = vids.map(vid => VID.encode(vid));
+            const decoded = encoded.map(vid => VID.decode(vid));
```
Comment (miniscruff) [inline tests/unit/versioning/VersionID.spec.js]:
```suggestion
            const encoded = vids.map(VID.encode);
            const decoded = encoded.map(VID.decode);
```
Does this work?

File: tests/unit/versioning/VersionID.spec.js [hunk 12]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
 
-    it('base62 version test with smaller part3 - 2', () => {
-        const vid = '98376906950000099999R1x';
-        const encoded = VID.base62Encode(vid);
-        const decoded = VID.base62Decode(encoded);
-        assert.strictEqual(vid, decoded);
-    });
+        it('sorted in reversed chronological and alphabetical order', () => {
+            for (let i = 0; i < count; i++) {
+                if (i !== 0) {
```
Comment (miniscruff) [inline tests/unit/versioning/VersionID.spec.js]: again just start at 1

**Review by miniscruff — CHANGES_REQUESTED** [review]

File: lib/versioning/VersionID.js (hunk 8 above)
Comment (watsaqat) [reply lib/versioning/VersionID.js]: we're not trying to find the last index with null but the first index with null. TrimEnd doesn't work unfortunately.

Comment (miniscruff) [reply lib/versioning/VersionID.js]: oh well the first index would just be `indexOf` unless you specifically needed to start from the end for some reason.

File: lib/versioning/VersionID.js (hunk 7 above)
Comment (watsaqat) [reply lib/versioning/VersionID.js:125]: I wrote it this way to be explicit about what's happening and to avoid a negative predicate check. I agree it looks weird, but I feel the behavior is obfuscated the other way.

File: tests/unit/versioning/VersionID.spec.js (hunk 10 above)
Comment (watsaqat) [reply tests/unit/versioning/VersionID.spec.js:37]: I didn't write this test originally, just moved it around, so I'd rather leave it in case the original author wanted to revisit it/ modify it. I agree we probably don't need it but whoever added the skip intentionally didn't delete it and skipped it instead.

File: tests/unit/versioning/VersionID.spec.js (hunk 9 above)
Comment (watsaqat) [reply tests/unit/versioning/VersionID.spec.js]: with branch prediction this is going to be barely noticeable 
https://measurethat.net/Benchmarks/Show/18409/0/branch-prediction
I don't see the need for micro-optimizing unit tests in general, but that's just my opinion.

File: tests/unit/versioning/VersionID.spec.js (hunk 11 above)
Comment (watsaqat) [reply tests/unit/versioning/VersionID.spec.js]: I'm trying to find the es-lint rule for this. Do you have a link? We don't enforce it right now though and I'm not sure if we should be introducing linting rules in a random pr.

File: lib/versioning/VersionID.js (hunk 6 above)
Comment (watsaqat) [reply lib/versioning/VersionID.js]: oh I was thinking in case we add a 4th field in the future and use it a separator or something. But we would just use the offsets so just disregard that.

File: lib/versioning/VersionID.js (hunk 8 above)
Comment (watsaqat) [reply lib/versioning/VersionID.js]: there's a comment above with the benchmarks.

Comment (watsaqat) [reply lib/versioning/VersionID.js]: I originally did it with just regex but Vianney suggested to optimize

File: lib/versioning/VersionID.js [hunk 13]
```diff
@@ -181,9 +188,15 @@ function base62Encode(str) {
...
-        enc3;
+    const padded1 = (B62V_EPAD + enc1).slice(-B62V_EPAD.length);
+    const padded2 = (B62V_EPAD + enc2).slice(-B62V_EPAD.length);
+    const enc3String = enc3.toString();
+    const totalLength = padded1.toString().length +
+        padded2.toString().length +
+        enc3String.length;
+    const enc3PaddingNeeded = 32 - totalLength + enc3String.length;
+
+    return padded1 + padded2 + enc3.toString().padEnd(enc3PaddingNeeded, '\u0000');
```
Comment (jonathan-gramain) [inline lib/versioning/VersionID.js]: I don't think we can return null bytes in the encoded version: it should only contain base62 characters. Even if we are allowed other characters than base62 (letters and digits) we have to consider what the application would do with such characters, it's surely calling for issues to include nonprintable characters in the exposed version ID.

**Review by jonathan-gramain — CHANGES_REQUESTED** [review]

File: tests/unit/versioning/VersionID.spec.js (hunk 11 above)
Comment (miniscruff) [reply tests/unit/versioning/VersionID.spec.js]: This is a linting rule? I was just suggesting it to pass the function directly instead of passing an inline function for the same thing. Not a big deal either way.

File: tests/unit/versioning/VersionID.spec.js (hunk 9 above)
Comment (miniscruff) [reply tests/unit/versioning/VersionID.spec.js]: My comment was not about optimization it was about reducing a tab and readability. I rarely try and optimize tests unless its like super obvious or less code in general.

File: lib/versioning/VersionID.js (hunk 13 above)
Comment (watsaqat) [reply lib/versioning/VersionID.js:199]: I think the only way then is to rely on the fact that the unpadded string should be 22 bytes and then add noise. Then decoding is a matter of stripping the last 10 bytes

File: lib/versioning/VersionID.js (hunk 8 above)
Comment (miniscruff) [reply lib/versioning/VersionID.js]: oh yea missed it, sorry

File: tests/unit/versioning/VersionID.spec.js (hunk 10 above)
Comment (miniscruff) [reply tests/unit/versioning/VersionID.spec.js:37]: hmm fair, maybe we can propose someone else look at it. It will just stay around indefinitely otherwise is my guess.

File: lib/versioning/VersionID.js (hunk 13 above)
Comment (watsaqat) [reply lib/versioning/VersionID.js]: @vrancurel thoughts on this?

Comment (vrancurel) [reply lib/versioning/VersionID.js]: oh right I though we were padding null bytes in the decoded version. Yes the encoded key shall only return base62 characters.

**Comment by watsaqat** [comment]
putting back in draft

**Review by vrancurel — APPROVED** [review]

File: lib/versioning/VersionID.js [hunk 14]
```diff
@@ -202,8 +211,19 @@ function base62Decode(str) {
...
+        // take off prefix 0s which represent null bytes
+        let idx = 0;
+        while (idx < enc3.length) {
+            if (enc3[idx] === '0') {
+                idx++;
+            } else {
+                break;
+            }
+        }
+        enc3 = enc3.slice(idx);
```
Comment (jonathan-gramain) [inline lib/versioning/VersionID.js:225]: In JS this should be relatively efficient and simplifies code a bit:

```suggestion
        enc3 = enc3.replace(/^0*/, '');
```

File: tests/unit/versioning/VersionID.spec.js (hunk 12 above)
Comment (jonathan-gramain) [inline tests/unit/versioning/VersionID.spec.js]:
```suggestion
            for (let i = 1; i < count; i++) {
```

File: tests/unit/versioning/VersionID.spec.js [hunk 15]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
-    it('should encode and decode base62 versionIds', () => {
-        const encoded = vids.map(vid => VID.base62Encode(vid));
-        const decoded = encoded.map(vid => VID.base62Decode(vid));
-        assert.strictEqual(vids.length, count);
-        assert.deepStrictEqual(vids, decoded);
+        it('should encode and decode correctly with new 32 byte format', () => {
+            const encoded = vids.map(vid => VID.encode(vid));
+            const decoded = encoded.map(vid => VID.decode(vid));
+            assert.strictEqual(vids.every(x => x.length === 27), true);
+            assert.strictEqual(encoded.every(x => x.length === 32), true);
```
Comment (jonathan-gramain) [inline tests/unit/versioning/VersionID.spec.js]:
```suggestion
            assert(vids.every(x => x.length === 27));
            assert(encoded.every(x => x.length === 32));
```

File: lib/versioning/VersionID.js [hunk 16]
```diff
@@ -240,14 +260,19 @@ function encode(str, encType = ENC_TYPE_HEX) {
  * @return {(string|Error)} - the decoded versionId or an error
  */
 function decode(str) {
-    if (str.length < 40) {
+    // default format is exactly 32 characters when encoded
+    if (str.length === 32) {
         return base62Decode(str);
```
Comment (jonathan-gramain) [inline lib/versioning/VersionID.js]: I'd suggest adding an assert to check the decoded size, because for such an important invariant, in the unlikely case it would be broken at some point it seems better to crash than to store invalid version IDs with the risk of corruption:

```suggestion
        const decoded = base62Decode(str);
        assert.strictEqual(decoded.length, 27);
        return decoded;
```

File: tests/unit/versioning/VersionID.spec.js [hunk 17]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
 }
 
-describe('test generating versionIds', () => {
-    const count = 1000;
+function generateRandomVIDs(count) {
     const vids = Array(count).fill(null);
     for (let i = 0; i < count; i++) {
         vids[i] = VID.generateVersionId(randkey(15), 'PARIS');
     }
     process.env.VID_CRYPTO_PASSWORD = randkey(64);
```
Comment (jonathan-gramain) [inline tests/unit/versioning/VersionID.spec.js]: I'd put this line out of the function since it does not look related

File: tests/unit/versioning/VersionID.spec.js [hunk 18]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
+        env.S3_VERSION_ID_ENCODING_TYPE = 'base62';
+        const vids = generateRandomVIDs(count);
 
-    it('base62 version test with smaller part3 - 2', () => {
-        const vid = '98376906950000099999R1x';
-        const encoded = VID.base62Encode(vid);
-        const decoded = VID.base62Decode(encoded);
-        assert.strictEqual(vid, decoded);
-    });
+        it('sorted in reversed chronological and alphabetical order', () => {
```
Comment (jonathan-gramain) [inline tests/unit/versioning/VersionID.spec.js:65]: Woulnd't this test better be done outside of the encoding scheme `describe` blocks, only once? Because I don't think it depends on the encoding scheme, it's just checking that internal version IDs are properly ordered without encoding/decoding anything.

File: tests/unit/versioning/VersionID.spec.js [hunk 19]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
     process.env.VID_CRYPTO_PASSWORD = randkey(64);
+    return vids;
+}
+
+const count = 1000000;
+
+describe('test generating versionIds', () => {
+    describe('legaxy hex encoding', () => {
+        env.S3_VERSION_ID_ENCODING_TYPE = 'hex';
+        const vids = generateRandomVIDs(count);
```
Comment (jonathan-gramain) [inline tests/unit/versioning/VersionID.spec.js:28]: Linked with my other comment, we may also move the vid generation outside of the encoding-specific test block, after `const count = 1000000`

File: tests/unit/versioning/VersionID.spec.js [hunk 20]
```diff
@@ -9,83 +10,124 @@ function randkey(length) {
...
-    it('should encode and decode versionIds', () => {
-        const encoded = vids.map(vid => VID.hexEncode(vid));
-        const decoded = encoded.map(vid => VID.hexDecode(vid));
-        assert.strictEqual(vids.length, count);
-        assert.deepStrictEqual(vids, decoded);
-    });
+        it('should encode and decode versionIds', () => {
+            const encoded = vids.map(vid => VID.hexEncode(vid));
+            const decoded = encoded.map(vid => VID.hexDecode(vid));
+            assert.strictEqual(vids.length, count);
```
Comment (jonathan-gramain) [inline tests/unit/versioning/VersionID.spec.js]: You could get rid of this line as it does not check anything more than the next line and only checks that the original `vids` array indeed has `count` versions in it (so testing the test case in a way)

**Review by jonathan-gramain — APPROVED** [review]
LGTM with some suggestions

File: tests/unit/versioning/VersionID.spec.js (hunk 19 above)
Comment (watsaqat) [reply tests/unit/versioning/VersionID.spec.js:28]: the env var `env.S3_VERSION_ID_ENCODING_TYPE = 'hex';` needs to be set before the run. Otherwise the info field gets chopped off by default.

File: tests/unit/versioning/VersionID.spec.js (hunk 18 above)
Comment (watsaqat) [reply tests/unit/versioning/VersionID.spec.js:65]: I think its better to explicitly show that the new format won't break ordering in the spec

Comment (watsaqat) [reply tests/unit/versioning/VersionID.spec.js:65]: its only obvious to us that ordering is preserved because we know exactly what its doing (just chopping off the info field if there is one) in the new format.

Comment (jonathan-gramain) [reply tests/unit/versioning/VersionID.spec.js:65]: Yes, it makes sense since it is removing the info, I did not get this at first.

File: lib/versioning/VersionID.js [hunk 21]
```diff
@@ -116,12 +117,19 @@ function generateVersionId(info, replicationGroupId) {
     lastSeq = (lastTimestamp === ts) ? lastSeq + 1 : 0;
     lastTimestamp = ts;
 
+    // if USE_HEX_VERSION_ID is set, info is used. By default, it is not used.
```
Comment (jonathan-gramain) [inline lib/versioning/VersionID.js]:
```suggestion
    // if S3_VERSION_ID_ENCODING_TYPE is "hex", info is used. By default, it is not used.
```

**Review by jonathan-gramain — APPROVED** [review]

File: lib/versioning/VersionID.js (hunk 14 above)
Comment (watsaqat) [reply lib/versioning/VersionID.js:225]: leaving as is for a bit of additional perf: https://measurethat.net/Benchmarks/Show/18405/0/regex-vs-slice

**Review by miniscruff — APPROVED** [review]
approving just curious on one point
